lib/dao.js

```diff
@@ -509,6 +509,41 @@ DataAccessObject.upsert = function(data, options, cb) {
   if (id === undefined || id === null) {
     return this.create(data, options, cb);
   }
+  var doValidate = undefined;
+  if (options.validate === undefined) {
+    if (Model.settings.validateUpsert === undefined) {
+      if (Model.settings.automaticValidation !== undefined) {
+        doValidate = Model.settings.automaticValidation;
+      }
+    } else {
+      doValidate = Model.settings.validateUpsert;
+    }
+  } else {
+    doValidate = options.validate;
+  }
+
+  var forceId = this.settings.forceId;
+  if (forceId) {
+    options = Object.create(options);
+    options.validate = !!doValidate;
+    if (doValidate) {
+      Model.findById(id, options, function(err, model) {
+        if (err || !model) return cb(err || errorModelNotFound(id));
+        model.updateAttributes(data, options, cb);
+      });
+    } else {
+      const model = new Model({id: id}, {persisted: true});
+      model.updateAttributes(data, options, cb);
+    }
+    return;
+  }
+
+  function errorModelNotFound(idValue) {
```

Comment:
Isn't there a similar method already defined somewhere in this huge file, a method we can reuse? If it isn't, them please make one - i.e. move this function to the root level of this source file, so that other methods can use it too.